tools/src/main/java/org/apache/kafka/trogdor/coordinator/NodeManager.java

```diff
@@ -0,0 +1,262 @@
+                            log.info("{}: NodeManagerRunnable got InterruptedException", node.name());
+                            Thread.currentThread().interrupt();
+                        }
+                        if (fault == null) {
+                            if (!faultQueue.isEmpty()) {
+                                fault = faultQueue.remove(0);
+                            }
+                        }
+                    } finally {
+                        lock.unlock();
+                    }
+                }
+            } catch (Throwable e) {
+                log.warn("{}: exiting NodeManagerRunnable with exception", node.name(), e);
+            } finally {
+            }
+        }
+    }
+
+    NodeManager(Time time, Node node) {
+        this.time = time;
+        this.node = node;
+        this.client = new AgentClient(node.hostname(), Node.Util.getTrogdorAgentPort(node));
+        this.heartbeatMs = Node.Util.getIntConfig(node,
+                Platform.Config.TROGDOR_COORDINATOR_HEARTBEAT_MS,
+                Platform.Config.TROGDOR_COORDINATOR_HEARTBEAT_MS_DEFAULT);
+        this.runnable = new NodeManagerRunnable();
+        this.thread = new KafkaThread("NodeManagerThread(" + node.name() + ")", runnable, false);
+        this.thread.start();
+    }
+
+    private boolean sendFault(long now, Fault fault) {
+        try {
+            client.putFault(new CreateAgentFaultRequest(fault.id(), fault.spec()));
+        } catch (Exception e) {
+            log.warn("{}: error sending fault to {}.", node.name(), client.target(), e);
+            return false;
+        }
+        lock.lock();
+        try {
+            lastContactMs = now;
+        } finally {
+            lock.unlock();
+        }
+        return true;
+    }
+
+    private void sendHeartbeat(long now) {
+        AgentStatusResponse status = null;
+        try {
+            status = client.getStatus();
+        } catch (Exception e) {
+            log.warn("{}: error sending heartbeat to {}.", node.name(), client.target(), e);
+            return;
+        }
+        lock.lock();
+        try {
+            lastContactMs = now;
+        } finally {
+            lock.unlock();
+        }
+        log.debug("{}: got heartbeat status {}.", node.name(), status);
+    }
+
+    public void beginShutdown() {
+        lock.lock();
+        try {
+            if (shutdown)
+                return;
+            log.trace("{}: beginning shutdown.", node.name());
+            shutdown = true;
+            cond.signalAll();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public void waitForShutdown() {
+        log.trace("waiting for NodeManager({}) shutdown.", node.name());
+        try {
+            thread.join();
+        } catch (InterruptedException e) {
+            log.error("{}: Interrupted while waiting for thread shutdown", node.name(), e);
+            Thread.currentThread().interrupt();
+        }
+    }
```

Comment:
`beginShutdown()` and `waitForShutdown` are defined in `Agent`, `Coordinator` and `NodeManager`. You could potentially inherit common code for these threads.

Reply:
The locking is different in all the threads, though, so the code is different for each.